apps/crn-server/src/data-providers/contentful/working-groups.data-provider.ts

```diff
@@ -0,0 +1,261 @@
+    >(FETCH_WORKING_GROUPS, {
+      limit: take,
+      skip,
+      where,
+    });
+
+    if (!workingGroupsCollection?.items) {
+      return {
+        total: 0,
+        items: [],
+      };
+    }
+
+    return {
+      total: workingGroupsCollection?.total,
+      items: workingGroupsCollection?.items
+        .filter((x): x is WorkingGroupItem => x !== null)
+        .map(parseContentfulGraphQlWorkingGroup),
+    };
+  }
+
+  async fetchById(id: string): Promise<WorkingGroupDataObject | null> {
+    const { workingGroups } = await this.contentfulClient.request<
+      FetchWorkingGroupByIdQuery,
+      FetchWorkingGroupByIdQueryVariables
+    >(FETCH_WORKING_GROUP_BY_ID, { id });
+
+    if (!workingGroups) {
+      return null;
+    }
+
+    return parseContentfulGraphQlWorkingGroup(workingGroups);
+  }
+
+  async update(
+    id: string,
+    update: WorkingGroupUpdateDataObject,
+  ): Promise<void> {
+    const environment = await this.getRestClient();
+    const workingGroup = await environment.getEntry(id);
+
+    await Promise.all(
+      workingGroup.fields.deliverables['en-US'].map(
+        async (deliverable: Link<'Entry'>) => {
+          const deliverableId = deliverable.sys.id;
+          await contentfulRateLimiter.removeTokens(1);
+          const deliverableEntry = await environment.getEntry(deliverableId);
+          await contentfulRateLimiter.removeTokens(1);
+
+          if (deliverableEntry.isPublished()) {
+            await contentfulRateLimiter.removeTokens(1);
+            await deliverableEntry.unpublish();
+          }
+
+          await contentfulRateLimiter.removeTokens(1);
+          await deliverableEntry.delete();
+        },
+      ),
+    );
+
+    const publishedDeliverables = await Promise.all(
+      update.deliverables.map(async (deliverable) => {
+        await contentfulRateLimiter.removeTokens(1);
+        const entry = await environment.createEntry(
+          'workingGroupDeliverables',
+          {
+            fields: addLocaleToFields(deliverable),
+          },
+        );
+
+        await contentfulRateLimiter.removeTokens(1);
+        return entry.publish();
+      }),
+    );
+
+    const deliverables = publishedDeliverables.map((deliverable) => ({
+      sys: {
+        type: 'Link',
+        linkType: 'Entry',
+        id: deliverable.sys.id,
+      },
+    }));
+
+    await contentfulRateLimiter.removeTokens(2);
+    await patchAndPublish(workingGroup, { deliverables });
+  }
+}
+export const parseContentfulGraphQlWorkingGroup = (
+  item: WorkingGroupItem,
+): WorkingGroupDataObject => {
+  const {
+    title,
+    complete,
+    externalLink,
+    calendars,
+    deliverablesCollection,
+    sys,
+    shortText,
+    description,
+    membersCollection,
+  } = item;
+
+  const deliverables = (deliverablesCollection?.items || []).reduce(
+    (
+      wgDeliverables: WorkingGroupDataObject['deliverables'][number][],
+      deliverableItem,
+    ) => {
+      if (deliverableItem) {
```

Comment:
correct me if i am wrong but aren't you simply filtering out the nulls/undefined and mapping to a slighlty different structure? I would prefer a `.filter` and a `.map` even if its more costly etc, just cause its more readable. Your solution is obvs more optimised (reduce + push rather than spreading) but less readable.

Reply:
done here [review: refactor to use map+filter instead of reduce](https://github.com/yldio/asap-hub/pull/3110/commits/7ede2e49d24a6a2fc04350324a5dd6387c89536c)